slot.go

```diff
@@ -134,30 +176,75 @@ func (s *Slot) findKey(search []*pkcs11.Attribute) (*Key, error) {
 func describeKey(obj obj) (c, t string, err error) {
 	attrs, err := obj.ctx.GetAttributeValue(obj.h, obj.o, []*pkcs11.Attribute{
 		&pkcs11.Attribute{Type: pkcs11.CKA_CLASS},
-		&pkcs11.Attribute{Type: pkcs11.CKA_KEY_TYPE},
 	})
 	if err != nil {
+		err = fmt.Errorf("Error reading CKA_CLASS: %x", err)
 		return
 	}
-
 	switch btoi(attrs[0].Value) {
 	case pkcs11.CKO_PUBLIC_KEY:
 		c = "public"
 	case pkcs11.CKO_PRIVATE_KEY:
 		c = "private"
+	case pkcs11.CKO_CERTIFICATE:
+		c = "public"
+		certType, cerr := obj.ctx.GetAttributeValue(obj.h, obj.o, []*pkcs11.Attribute{
+			&pkcs11.Attribute{Type: pkcs11.CKA_CERTIFICATE_TYPE},
+		})
+		if cerr != nil {
+			err = fmt.Errorf("Error reading CKA_CERTIFICATE_TYPE: %x", cerr)
+			return
+		}
+		switch btoi(certType[0].Value) {
+		case pkcs11.CKC_X_509:
+			certValue, cerr := obj.ctx.GetAttributeValue(obj.h, obj.o, []*pkcs11.Attribute{
+				&pkcs11.Attribute{Type: pkcs11.CKA_VALUE},
+			})
+			if cerr != nil {
+				err = fmt.Errorf("Error reading CKA_VALUE: %x", cerr)
+				return
+			}
+			cert, cerr := x509.ParseCertificate(certValue[0].Value)
+			if cerr != nil {
+				err = fmt.Errorf("Error parsing x509 certificate: %x", cerr)
+				return
+			}
+			switch cert.PublicKeyAlgorithm {
+			case x509.RSA:
+				t = "x509:rsa"
```

Comment:
Yep; I didn’t know if there would be an easier way to check whether the material was x509-encoded.